source/Octopus.Cli/Commands/Deployment/DeploymentCommandBase.cs

```diff
@@ -81,11 +82,68 @@ protected override async Task ValidateParameters()
-
             if (IsTenantedDeployment && !await Repository.SupportsTenants().ConfigureAwait(false))
                 throw new CommandException("Your Octopus Server does not support tenants, which was introduced in Octopus 3.4. Please upgrade your Octopus Server, enable the multi-tenancy feature or remove the --tenant and --tenanttag arguments.");
+            /*
+             * A create release operation can also optionally deploy the release, however any invalid options that
+             * are specific only to the deployment will fail after the release has been created. This can leave
+             * a deployment in a half finished state, so this validation ensures that the input relating to the
+             * deployment is valid so missing or incorrect input doesn't stop stop the deployment after a release is
+             * created.
+             *
+             * Note that certain validations still need to be done on the server. Permissions and lifecycle progression
+             * still rely on server side validation.
+             */
+            
+            // Make sure the tags are valid
+            foreach (var tenantTag in TenantTags)
+            {
+                // Verify the format of the tag
+                var parts = tenantTag.Split(Seperator);
+                if (parts.Length != 2 || string.IsNullOrEmpty(parts[0]) || string.IsNullOrEmpty(parts[1]))
+                {
+                    throw new CommandException(
+                        $"Canonical Tag Name expected in the format of `TagSetName{Seperator}TagName`");
+                }
+                // Verify the presence of the tag
+                var tagSets = await Repository.TagSets.FindByName(parts[0]).ConfigureAwait(false);
+                if (tagSets?.Tags?.All(tag => parts[1] != tag.Name) ?? true)
+                {
+                    throw new CommandException(
+                        $"Unable to find matching tag from canonical tag name `{tenantTag}`");
+                }
+            }
+
+            // Make sure the tenants are valid
+            foreach (var tenantName in Tenants)
+            {
+                var tenant = await Repository.Tenants.FindByName(tenantName).ConfigureAwait(false);
+                if (tenant == null)
+                {
+                    throw new CommandException(
+                        $"Could not find the tenant {tenantName} on the Octopus Server");
+                }
+            } 
+            
+            // Make sure environment is valid
+            foreach (var environmentName in DeployToEnvironmentNames)
+            {
+                var environments = await Repository.Environments.FindByNames(DeployToEnvironmentNames).ConfigureAwait(false);
+                var missingEnvironment = DeployToEnvironmentNames
+                    .Where(env => environments.All(env2 => env2.Name != env))
```

Comment:
I am not sure about the impl of `All(predicate)`, but maybe it makes sense to call `All` outside the loop?

Reply:
Actually the outer loop was not required at all here, so it will be removed.